src/helpers/buffers/fsv.rs

```diff
@@ -38,27 +41,39 @@ impl<const N: usize> FixedSizeByteVec<N> {
         Self {
             data: vec![0_u8; N * capacity],
             added: bitvec![0; capacity],
-            taken: 0,
+            capacity,
+            end: capacity,
         }
     }
 
     /// Inserts a new element to the specified position, returning the previous element at this `index`.
     /// ## Panics
     /// Panics if `index` is out of bounds or if something was previously inserted at `index`.
-    pub fn insert(&mut self, index: usize, elem: &[u8; N]) {
-        // if index is out of bounds, this line will panic, there is no need for additional check
-        // TODO: save runtime cost with `debug_assert!()`, though that would affects panic tests
-        assert!(!self.added[index]);
+    pub fn insert<D: Debug>(&mut self, channel: D, index: usize, elem: &[u8; N]) {
+        // Translate from an absolute index into a relative one.
+        let index = index
+            .checked_sub(self.end - self.capacity)
+            .unwrap_or_else(|| panic!("Duplicate send for index {index} on channel {channel:?}"));
```

Comment:
doesn't it indicate that index is smaller than lower bound of accepted range?

Reply:
Yes.  But index is unsigned and so it cannot be below zero.  Any value below the start of the range but non-zero therefore must have already been taken and sent out, so it's a duplicate.